[Backend] Improve dot support to target FMA

binarman opened the pull request:
This PR:
- Refactors FMA dot implementation
- Supports dot3d in FMA path
- Fixes several issues in operand offset computation
- Enables small dot operands

This PR is a part of PR series. Final goal is to improve efficiency of small dot operations and bypass as much shared memory accesses as possible.

Rough list of PRs:
- [ ] Basic FMA dot fixes, dot 3d support and relaxing small dimensions for dot **(this PR)** #4516
- [x] Blocked->dotOp shared memory bypassing #4538
- [ ] Accelerate AMD Matmul + emit dot operations #4594
- [ ] Layout optimization, so operand B is loaded in proper mfma layout and do not need to go through LDS #4581 
- [ ] Vectorization optimization of dot operands/results (in case llvm can not do this internally)
- [ ] Reduction operation hoisting out of the K loop (reduction operation is a byproduct of layout optimization step) #4559

antiagainst commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,229 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
-  SmallVector<Value> aOff(aNumPtr);
-  for (int i = 0; i < aNumPtr; ++i) {
-    aOff[i] = add(mul(offA0, strideA0), mul(offA1, strideA1));
+/**
```
This is not the commonly used style. Can we follow the common style in the codebase to be consistent?

binarman replied:
I will change it.
Do we have a code style guide for stuff like this?

Sometimes it is not obvious which option to pick, I've seen few different styles of comments in many places in our code base and seems this depends on a code author taste.

antiagainst replied:
Yeah it's a bit confusing. I think we should follow MLIR style in general: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Operation.h

antiagainst commented on python/test/unit/language/test_core.py:
```diff
@@ -3282,6 +3286,9 @@ def kernel(X, stride_xm, stride_xk, Y, stride_yk, stride_yn, W, stride_wn, strid
         return
     # make sure ld/st are vectorized
     ptx = pgm.asm['ptx']
+    is_fma = K < 16 or N < 16 or M < 16
```
Can we skip testing these small matmuls for nvidia?

antiagainst commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,229 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
-  auto bTensorTy = cast<MemDescType>(B.getType());
-  auto bLayout = cast<SharedEncodingAttr>(bTensorTy.getEncoding());
-  auto bShapePerCTA = getShapePerCTA(bTensorTy);
+Value loadFMAOp(Value dotOp, Value llA, BlockedEncodingAttr dLayout,
```
There are lots of magic indexing and calculation in this function in general. Can you provide more comments inside to make it easier for others to follow?

antiagainst reviewed (changes requested):
First batch of comments; I still need to review `SharedToDotOperandFMA.cpp` more carefully.

binarman commented on lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/FMA.cpp:
```diff
@@ -39,61 +40,56 @@ LogicalResult convertFMADot(triton::DotOp op, triton::DotOp::Adaptor adaptor,
   auto aTensorTy = cast<RankedTensorType>(A.getType());
-  auto bTensorTy = cast<RankedTensorType>(B.getType());
   auto dTensorTy = cast<RankedTensorType>(D.getType());
+  auto dElemTy = dTensorTy.getElementType();
```
In this PR yes. It will be used in later parts to choose data specific intrinsics instead of simple FMA.
I will remove it here.

binarman commented on lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/FMA.cpp:
```diff
@@ -39,61 +40,56 @@ LogicalResult convertFMADot(triton::DotOp op, triton::DotOp::Adaptor adaptor,
+        unsigned idx[] = {b, m, n};
+        unsigned linearIdx = 0;
+        for (auto dim : llvm::reverse(order)) {
+          linearIdx = linearIdx * retSize[dim] + idx[dim];
```
We have values scattered across multiple dimensions of the tensor, but in LLVM IR Triton stores them in linear structure.

This part computes linear index in this structure where to put dot result according it's batch, M and N coordinates.

I will put this part in separate function.

binarman commented:
> When addressing comments, please make sure to add new commits and not squashing into existing ones. Otherwise it's hard to re-review again.

hmm, let me rework this with merge commits, I did not notice this comment in time

antiagainst commented on include/triton/Conversion/TritonGPUToLLVM/Utility.h:
```diff
@@ -1474,6 +1477,35 @@ inline bool isLayoutMmaV1(Attribute layout) {
 }
 
+/// @brief Extend 2d shared object to 3d.
+/// Generalizing dot operand shared to distributing conversion.
```
"Generalizing dot operand to shared distributing conversion"?

binarman replied:
Yep, this part does not sound great

antiagainst commented on third_party/nvidia/backend/compiler.py:
```diff
@@ -15,7 +15,16 @@
 def min_dot_size(target: GPUTarget):
-    return lambda lhsType, rhsType: (16, 32, 16) if lhsType.is_int8() else (16, 16, 16)
+
+    def fma_supported(lhsType, rhsType):
```
Yeah common path supporting it gives the capabilities. But whether to turn it on it's an explicit judgement call. We don't use NVIDIA code path; so it should be leaving to folks using it.

antiagainst commented on python/test/unit/language/test_core.py:
```diff
@@ -3092,13 +3092,21 @@ def convert_fp8_to_fp32(x, device, dtype_str):
 def test_dot(M, N, K, num_warps, col_a, col_b, epilogue, input_precision, in_dtype, out_dtype, kpack, num_ctas, device):
     if is_interpreter():
+        if M < 16 or N < 16 or K < 16:
+            pytest.skip("small dots are aupported only on HIP at the moment")
```
typo: supported. similarly for the following.

antiagainst commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+  unsigned nonK;
+};
+
+/// @brief put elements from Value vec to appropriate indexes in opValues array
```
Can you follow MLIR comment style (e.g., https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Operation.h) to write docs?

antiagainst reviewed (changes requested):
Thanks for improving the implementation! I still have a few nits. 

The change is quite intensive and I'm not super familiar with shared layout indexing and such. @zhanglx13 please take another look.

antiagainst reviewed (changes requested):
This now looks good to me. But please wait for @zhanglx13 to double check the logic there given it's fairly involved. I'll approve after Lixun approves.

zhanglx13 commented on include/triton/Dialect/TritonGPU/IR/Dialect.h:
```diff
@@ -152,6 +152,18 @@ void dumpHWLayout(RankedTensorType tensorType);
 std::string getLayoutStr(RankedTensorType tensorType, bool useHWPointOfView);
 
+template <typename T>
+llvm::SmallVector<T> expandMatrixShapeWithBatch(llvm::ArrayRef<T> s) {
```
Can you move the implementation to .cpp file?

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
-  auto bTensorTy = cast<MemDescType>(B.getType());
-  auto bLayout = cast<SharedEncodingAttr>(bTensorTy.getEncoding());
-  auto bShapePerCTA = getShapePerCTA(bTensorTy);
+Value loadFMAOp(Value dotOp, Value llA, BlockedEncodingAttr dLayout,
```
The 1st arg is the src of local_local. Can you change it to something like "srcVal"?

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
-  SmallVector<Value> aOff(aNumPtr);
-  for (int i = 0; i < aNumPtr; ++i) {
-    aOff[i] = add(mul(offA0, strideA0), mul(offA1, strideA1));
+struct DimNumbers {
```
Maybe call it "DimIdx" given it's the indices of different dims.

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+  auto sharedOrder = expandMatrixOrderWithBatch(sharedLayout.getOrder());
+  // compute contiguity of fastest dimension in shared layout.
+  unsigned vectorSize = sizePerThread[sharedOrder[0]];
+  if (sharedLayout.getMaxPhase() > 1)
```
Can you use isSwizzled() to make it cleaner?

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+///
+/// first_element_offset depends on lane Id and warp Id
+/// constant_offset depends on value number, which is same for all threads.
+/// \returns first_element_offset
```
"... is computed with formula"
"constant_offset depends on value index, ..."

Is it assumed that elements for the same thread are contiguous in LDS?

binarman replied:
No, it only assumes that there are a fixed stride for every value in thread, which is same for every threads.

Consider an example with three threads(threads 0, 3, 32), mfma 32 dot operand A with kwidth=2.

Value number 0 corresponds to matrix elements:
-  [0, 0] for thread 0
-  [3, 0] for thread 3
-  [0, 2] for thread 32

Value number 2 corresponds to elements:
- [0, 4] for thread 0
- [3, 4] for thread 3
- [0, 6] for thread 32

Multidimensional offsets between 2 and 0 element stored in every thread are following:
- thread 0: [0, 4] - [0, 0] = [0, 4]
- thread 3: [3, 4] - [3, 0] = [0, 4]
- thread 32: [0, 6] - [0, 2] = [0, 4]

So in case memory is not swizzled and plain memory offset depends on multidimensional element coordinates lineary, we can split "thread dependant" part of offset and fixed offset for every value.

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+  // If memory is not swizzled, algorithm breaks element offset pointer into
+  // constant and non-constant part. Non-constant (depends on thread id) part is
+  // same for all elements, so it is computed only once. basePtr stores this
+  // non-constant part
```
"Non-constant (depends on thread id) part is same for all elements,"
->
"Non-constant (depends on thread id) part is the offset of the first element of the thread, which is same for all elements of the thread."

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+  // This loop nest iterates over all values loaded in one thread across batch,
+  // k and nonK dimensions. Blocked dot operand layout allocates data in tiles
+  // of size <sizePerThread>*<threadsPerWarp>*<numberWarps> for batch and nonK
+  // dimensions, if tensor shape is larger than tile, pattern repeats. To take
```
"dimensions, if tensor shape is larger.." --> "dimensions. If tensor shape is larger..."

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+
+  auto sharedOrder = expandMatrixOrderWithBatch(sharedLayout.getOrder());
+  // compute contiguity of fastest dimension in shared layout.
+  unsigned vectorSize = sizePerThread[sharedOrder[0]];
```
Let's say for opIdx=0, and K = 16. This will lead to vectorSize = 16, which is larger than the max vectorSize of a single ds_read instruction. 
In the current implementation, K=16 will go to the mfma path. But in the future, we want to to try fma for small-M but larger K. So it's better to restrict vectorSize <= 128B/elemSize

binarman replied:
This is a common code base. Let's think about this in next PRs.
Maybe add a new functions "getMaxSharedLoadWidth" function in target info?

zhanglx13 commented on include/triton/Conversion/TritonGPUToLLVM/Utility.h:
```diff
@@ -1474,6 +1477,37 @@ inline bool isLayoutMmaV1(Attribute layout) {
+/// \param shape shape of a tensor represented by smemObj
+/// \returns shared object describing 3d tensor
+inline SharedMemoryObject
+getExpandedSharedMemoryObject(ConversionPatternRewriter &rewriter, Location loc,
```
Is there a reason to keep this impl in .h file?

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/FMA.cpp:
```diff
@@ -1,28 +1,42 @@
+    for (auto dim : order) {
+      spatialIdx[dim] = curIdx % perTileShape[dim];
+      curIdx /= perTileShape[dim];
+    }
```
Is this logic the same as delinearize()?

binarman replied:
Yes, there were no function with integer arguments, but I will add it

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/FMA.cpp:
```diff
@@ -34,68 +48,60 @@ LogicalResult convertFMADot(triton::DotOp op, triton::DotOp::Adaptor adaptor,
-      getValueTableFromStructFMA(llA, K, M, mShapePerCTATile, mSizePerThread,
-                                 rewriter, loc, typeConverter, aTensorTy);
+      getValueTableFromStructFMA(llA, {perThreadShape[0], perThreadShape[1], K},
+                                 2, 1, rewriter, loc, order);
```
Can you add some inline comment for kDim and nonKDim like 
/*kDim*/2, /*nonKDim*/1

zhanglx13 commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp:
```diff
@@ -71,164 +33,338 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+///
+/// This function maps elements of 3d sub-tensor in linear array.
+/// Axes are arranged in following order from fastest to slowest: [nonKdim,
+/// kDim, bDim]
```
why does nonKdim "faster" than the kDim? Or is this just the default setting?

binarman replied:
Sorry, I forgot to change this comment. It is outdated, in this version order is taken from `opOrder` parameter and could be arbitrary.

For now this argument is inherited from parent blocked layout, but in general we can use different "register" order (warp, and lane orders should be the same as parent blocked layout).

Let me try this in a followup PR, I would like to make some performance measurements as well.

binarman commented on third_party/amd/backend/compiler.py:
```diff
@@ -13,16 +13,30 @@
-    # Other architectures will only support 16,16,16
-    return lambda lhsType, rhsType: (16, 16, 16)
+        return get_gfx9_limits
+    # Other architectures will only support 16,16,16 with mfma instructions
```
Right, will change this comment

binarman commented on python/test/unit/language/test_core.py:
```diff
@@ -3683,6 +3700,8 @@ def kernel(
         out = numpy_random((B, M, N), dtype_str='int32', rs=rs)
     else:
+        x *= 0.1
+        y *= 0.1
```
Precision is not enough for some fp16xfp16->fp16 tests. Without this and FMA path I get errors like this:
```
FAILED python/test/unit/language/test_core.py::test_dot3d[1-1-64-64-64-32-32-float16-float16] - AssertionError: 
Not equal to tolerance rtol=0.01, atol=0.01

Mismatched elements: 9 / 4096 (0.22%)
Max absolute difference: 0.0781
Max relative difference: 0.833
 x: array([[[ -2.621 ,  10.03  , -13.05  , ...,   9.05  , -24.95  ,
          16.52  ],
        [-13.04  ,  15.22  , -11.414 , ...,   6.58  , -14.41  ,...
 y: array([[[ -2.627 ,  10.04  , -13.03  , ...,   9.05  , -24.97  ,
          16.48  ],
        [-13.02  ,  15.21  , -11.4   , ...,   6.562 , -14.414 ,...

```

By making gemm arguments smaller we effectively make absolute tolerance larger.
I copied this trick from test_dot.

I am sure this is not a compiler issue, because I've implemented gemm using bunch of numpy data casts and strict accumulation order to emulate code generated by Triton to generate reference output and got no mismatch errors:

``` python
  if out_dtype_str == 'float16':
    for b in range(B):
        for m in range(M):
            for n in range(N):
                out_ref[b, m, n] = 0.0
                for k in range(K):
                    out_ref[b, m, n] = np.float32(out_ref[b, m, n]) + np.float32(x[b, m, k]) * np.float32(y[b, k, n])
```

Results are still not bitwise equal, but it seems the difference somewhere in the last 1-2 bits now.

binarman replied:
I think Nvidia backend could be affected by this problem as well(not yet, because for Nvidia is disabled in this PR), since we loose precision with FMA a lot. Unfortunately I can not check it at the moment, but will update it in a follow up PRs.

antiagainst replied:
I think we can still gate it behind hip for now to be conservative. We can broaden it later if we enable the nvidia path. I pushed 4db0218 for this.

pawelszczerbuk commented on lib/Dialect/TritonGPU/IR/Dialect.cpp:
```diff
@@ -3730,6 +3727,36 @@ std::string getDistributedLayoutStr(RankedTensorType tensorType,
 
+template <typename T>
+llvm::SmallVector<T>
+mlir::triton::gpu::expandMatrixShapeWithBatch(llvm::ArrayRef<T> s) {
```
Nit: here and below: "Tensor" shape instead of "Matrix"?

binarman commented on python/test/unit/language/test_core.py:
```diff
@@ -3689,6 +3706,10 @@ def kernel(
     if in_dtype_str == 'int8':
         out = numpy_random((B, M, N), dtype_str='int32', rs=rs)
     else:
+        if (BLOCK_M < 16 or BLOCK_N < 16) and out_dtype_str == 'float16':
+            # float16 accumulator in FMA dot loose precision too fast
+            x *= 0.1
+            y *= 0.1
```
Changed code and lost previous conversation
FYI: @pawelszczerbuk

pawelszczerbuk reviewed (approved):
Thank you!
